net: l2: ieee802154: mgmt: allow beacons without association bit

gumulka opened the pull request:
The Association permit bit shall be set to zero if the coordinator does not accept association requests.
Not accepting association request ist not a reason to filter the beacons from this coordinator during network scan. It is still a network, just one you cannot associate with.

gumulka committed:
net: l2: ieee802154: mgmt: allow beacons without association bit

The Association permit bit shall be set to zero if the coordinator does
not accept association requests.
Not accepting association request ist not a reason to filter the
beacons from this coordinator during network scan. It is still a
network, just one you cannot associate with.

Signed-off-by: Fabian Pflug <fabian.pflug@grandcentrix.net>

fgrandel commented:
Thanks, Fabian, for this PR. I think you're right, that even those networks should reported that one cannot associate to. But then it is not enough to just not drop them. I think we should at least:
* Report the flag in the shell, see https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/net/l2/ieee802154/ieee802154_shell.c#L216
* Update the corresponding unit test to ensure the flag is properly reported. You should test once with and once w/o the flag set. You can expand the test_scan unit test in ieee802154_shell_test.c but please don't just duplicate the test but keep it DRY. If you're on it, you could also test with and w/o beacon payload to ensure that your new feature is also fully covered.

fgrandel reviewed (changes requested):
see comment

gumulka commented:
>     * Report the flag in the shell, see https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/net/l2/ieee802154/ieee802154_shell.c#L216

Except, that this is not part of the shell, but part of the ieee802154 management layer. To report it in the shell, you would need to copy the bit to the scan context and read out the scan context from the shell.

I could log the information, but it would probably be best if this is DEBUG log level because it is not needed normally and will clutter the log.

With #76190 in mind, you could replace the `payload` field with a pointer to the pkg, so that the notified layers could access all information from the package, but then you probably don't need most other fields in the ieee802154_req_params struct and would have to do a major rewrite.

What is the way you want to go with it?

fgrandel commented:
> Except, that this is not part of the shell, but part of the ieee802154 management layer. To report it in the shell, you would need to copy the bit to the scan context and read out the scan context from the shell.

Yes, that's what I had in mind, similarly to what you did with the beacon payload.

Probably this pattern doesn't scale well in the long run as a lot more fields may be added in the future. But let's not overthink it for now as I'll be able to refactor that scanning process once it's getting out of hand.

fgrandel commented:
> you could replace the `payload` field with a pointer to the pkg

No, I wouldn't do that because it would require clients to know too much about the internal structure of packages which would break encapsulation.

gumulka committed:
net: l2: ieee802154: decrease time for testcase

When trying to implement a testcase for the previous commit, an error
occoured, that stopped the test from executing.

As the tests require a full IPv6 stack, the usual router solicitation
(RS) messages will be scheduled by the net stack. To avoid conflict
with RS messages the active scan test must be kept shorter than one
second, otherwise a race condition with additional packages being
reported by the fake driver might occur.

Signed-off-by: Fabian Pflug <fabian.pflug@grandcentrix.net>

gumulka committed:
net: l2: ieee802154: test association permit bit

Beacon frames can be received with association permit bit set or not,
but the correct information should be in the data given to the
application.

Signed-off-by: Fabian Pflug <fabian.pflug@grandcentrix.net>

fgrandel commented on include/zephyr/net/ieee802154_mgmt.h:
```diff
@@ -330,6 +330,8 @@ struct ieee802154_req_params {
 	/** Link quality information, between 0 and 255 */
 	uint8_t lqi;
+	/** Flag if association is allowed */
+	bool association_allowed;
```
Can we call this "assoc_permit", "association_permit" or "association_permitted" please (whatever you prefer), that's how the standard calls it, see section 8.4.3.1, table 8-94 and many other places in which "association permitted" is used.

fgrandel commented on subsys/net/l2/ieee802154/ieee802154_mgmt.c:
```diff
@@ -44,9 +44,7 @@ enum net_verdict ieee802154_handle_beacon(struct net_if *iface,
-	if (!mpdu->beacon->sf.association) {
-		return NET_DROP;
-	}
+	ctx->scan_ctx->association_allowed = mpdu->beacon->sf.association;
```
While you're on it you could (strictly optional) rename sf.association to sf.<whatever you chose above> for improved readability and consistency.

fgrandel commented on subsys/net/l2/ieee802154/ieee802154_shell.c:
```diff
@@ -214,9 +214,10 @@ static void scan_result_cb(struct net_mgmt_event_callback *cb,
-		      "LQI: %u\n", params.channel, params.pan_id,
-		      print_coordinator_address(buf, sizeof(buf)), params.lqi);
+		      "Channel: %u\tPAN ID: %u\tCoordinator Address: %s\t "
+		      "LQI: %u Associable: %s\n", params.channel, params.pan_id,
```
I personally would prefer "association permitted" for improved consistency and standard compliance (strictly optional).

gumulka replied:
I disliked it, because it is a long name and I don't want to clutter the shell to much, as it is already a pretty long prompt.

fgrandel commented on tests/net/ieee802154/l2/src/ieee802154_shell_test.c:
```diff
@@ -30,6 +30,8 @@ static struct net_if *net_iface;
 static struct net_mgmt_event_callback scan_cb;
 K_SEM_DEFINE(scan_lock, 0, 1);
 
+static bool expected_association_bit;
```
`expected_<however you named it>_bit`

fgrandel commented on tests/net/ieee802154/l2/src/ieee802154_shell_test.c:
```diff
@@ -126,8 +130,10 @@ static void test_scan_shell_cmd(void)
 
 	/* The beacon placed into the RX queue will be received and handled as
 	 * soon as this command yields waiting for beacons.
+	 * Scan should be as short as possible, because after 1 second the IPv6
```
```suggestion
	 * Scan should be as short as possible, because after 1 second an IPv6
```

fgrandel commented on tests/net/ieee802154/l2/src/ieee802154_shell_test.c:
Cannot comment on commits, therefore commenting here: I don't think the full test log needs to be added to the commit message. It would be more informative to have something like "As the tests require a full IPv6 stack, the usual router solicitation (RS) messages will be scheduled by the net stack. To avoid conflict with RS messages the active scan test must be kept shorter than one second, otherwise a race condition with additional packages being reported by the fake driver might occur".

fgrandel commented on tests/net/ieee802154/l2/src/ieee802154_shell_test.c:
```diff
@@ -62,6 +64,8 @@ static void scan_result_cb(struct net_mgmt_event_callback *cb, uint32_t mgmt_eve
 	zassert_equal(scan_ctx->lqi, EXPECTED_COORDINATOR_LQI,
 		      "Scan did not receive correct link quality indicator.");
+	zassert_equal(scan_ctx->association_allowed, expected_association_bit,
+		      "Scan did not set the association bit correctly.");
```
```suggestion
		      "Scan did not set the association permit bit correctly.");
```

This is not optional, as the standard clearly names is as such, see section 7.3.1.4, figure 7-7.

fgrandel commented on tests/net/ieee802154/l2/src/ieee802154_shell_test.c:
```diff
@@ -203,27 +231,26 @@ ZTEST(ieee802154_l2_shell, test_active_scan)
+	beacon_pkt[14] = 0x40;
+	expected_association_bit = false;
+
+	if (create_and_receive_packet(beacon_pkt, sizeof(beacon_pkt)) < 0) {
```
Well done, this is really sweet and short and easy to read. :-)

gumulka replied:
Thanks :)

fgrandel reviewed (changes requested):
Again mostly (optional) nitpicks except for minor issues.

Nicely split up into easy-to-review commits and well documented. It's fun to work with you.

fgrandel reviewed (approved):
LGTM. Thanks for your contribution!

fgrandel commented:
@rlubos Could you give this a second look to see whether it could be merged?

rlubos reviewed (dismissed):
Please rebase

aescolar commented:
> @aescolar is there a reason my tests were cancelled?

Sorry for that: https://discord.com/channels/720317445772017664/733037890514321419/1267759879460753438

fgrandel commented:
> Sorry for that

Np. tests manually restarted. Thanks for scheduling the fix.
